### ISSUE TITLE
Deadlock happens when allocating pool of threads and instantiating wasm in a worker

### ISSUE THREAD

**Issue by ravisumit33** [emscripten-core/emscripten#15100]
Below is a sample code using boost thread pool:

```
#include <boost/asio/post.hpp>
#include <boost/asio/thread_pool.hpp>
#include <boost/asio/use_future.hpp>


int task_num = 1;

int main()
{
    std::printf("Inside main\n");
    boost::asio::thread_pool tp(5);
    auto task = [](){
        std::stringstream msg;
        msg << "Task " << task_num++ << std::endl;
        std::cout << msg.str(); 
    };
    std::vector<std::future<void>> vec;
    for (int i=0; i < 5; ++i) {
        vec.push_back(boost::asio::post(tp, boost::asio::use_future(task)));
    }
    for(auto& fut: vec){
        printf("Before wait\n");
        fut.wait();
    }
    std::cout << "Completed" << std::endl;
    return 0;
}
```

I have observed few cases and they are confusing me:

- Using `-sPTHREAD_POOL_SIZE=16`
     - When I instantiate wasm on main thread (basically using generated html by em++) no deadlock happens but I get below warning on console.
     `boost.html:1246 Blocking on the main thread is very dangerous`
    - When I instantiate wasm on a worker, deadlock happens. I tried to debug and execution stalls after first `wait`. Function given to threads i.e. `task` also get executed but somehow wait is not resolved.

- Using `-sPROXY_TO_PTHREAD`
    - Both cases work and no deadlock happens.
    
What I was excepting is `-sPTHREAD_POOL_SIZE=16` to work also in the case when I am instantiating wasm in a worker since I am pre-allocating required number of threads.

**Comment by sbc100** [comment]
If `-sPROXY_TO_PTHREAD` works is there any reason you can't use that?

**Comment by ravisumit33** [comment]
@sbc100 `-sPROXY_TO_PTHREAD` doesn't work when you have to call functions other than main directly from javascript and they can also create & join threads. Related https://github.com/emscripten-core/emscripten/issues/15102

**Comment by kripken** [comment]
@sbc100 Do you mean testing of a pthreads application that is run entirely inside a worker? `test_pthreads_started_in_worker` does that.

(If I misunderstood, what does "thread 0" mean in your question?)

**Comment by sbc100** [comment]
Yes I was looking for something like `test_pthreads_started_in_worker`.. 

@ravisumit33 it seems like `test_pthreads_started_in_worker` does do some basic testing of running a pthread-using application in a worker.  Perhaps we need more coverage to catch the deadlock case you are encountering?   If you have any time to look into the cause of the dead lock and/or extending our test coverage any PRs would be very welcome.

@kripken .. I wonder if we can run `test_pthreads_started_in_worker` under node too?

**Comment by kripken** [comment]
@sbc100 Good question - I haven't tried that. Would it help here, though - this issue seems to not be about node in particular?

**Comment by sbc100** [comment]
No, it wouldn't help directly here.. I just dislike having things coverage that only exists in browser tests (unless its a browser-only feature which perhaps it could be argued this is.. although I node coverage of this would catch a lot of the same issues).
